prov/gni/src/gnix_msg.c

```diff
@@ -1050,127 +1038,85 @@ smsg_callback_fn_t gnix_ep_smsg_callbacks[] = {
 };
 
-#define GNIX_TAGGED_PCD_COMPLETION_FLAGS	(FI_MSG | FI_RECV | FI_TAGGED)
-
-static int __gnix_peek_request(struct gnix_fid_ep *ep,
-		struct gnix_fab_req *req,
-		void *addr,
-		size_t len,
-		void *context,
-		uint64_t tag)
+static int __gnix_peek_request(struct gnix_fab_req *req)
 {
-	struct gnix_fid_cq *recv_cq = ep->recv_cq;
+	struct gnix_fid_cq *recv_cq = req->gnix_ep->recv_cq;
 	int rendezvous = !!(req->msg.send_flags & GNIX_MSG_RENDEZVOUS);
-	void *peek_addr = addr;
-
-	/* all claim work is performed by the tag storage,
-	 * so nothing special here
-	 *
-	 * if no CQ, no data is to be returned. Just inform the user that a
-	 * message is present.
-	 */
+	int ret;
+
+	/* All claim work is performed by the tag storage, so nothing special
+	 * here.  If no CQ, no data is to be returned.  Just inform the user
+	 * that a message is present. */
 	GNIX_INFO(FI_LOG_EP_DATA, "peeking req=%p\n", req);
 	if (!recv_cq)
 		return FI_SUCCESS;
 
-	/* rendezvous messages on the unexpected queue won't have data
-	 *
-	 * additionally, if the cq format doesn't support passing a buffer
-	 * location and length, then data will not be copied
-	 * */
-	if (!rendezvous && peek_addr &&
-			!INVALID_PEEK_FORMAT(recv_cq->attr.format))
-		memcpy(peek_addr, (void *) req->msg.recv_addr,
-				len);
-	else
-		peek_addr = NULL;
+	/* Rendezvous messages on the unexpected queue won't have data.
+	 * Additionally, if the CQ format doesn't support passing a buffer
+	 * location and length, then data will not be copied. */
+	if (!rendezvous && req->msg.recv_addr &&
+	    !INVALID_PEEK_FORMAT(recv_cq->attr.format)) {
+		int copy_len = MIN(req->msg.send_len, req->msg.recv_len);
 
-	return __recv_completion(ep, req, context,
-				 GNIX_TAGGED_PCD_COMPLETION_FLAGS,
-				 len, peek_addr, req->msg.imm, tag,
-				 _gnix_vc_peer_fi_addr(req->vc));
+		memcpy((void *)req->msg.recv_addr, (void *)req->msg.send_addr,
+		       copy_len);
+	} else {
+		/* The CQE should not contain a valid buffer. */
+		req->msg.recv_addr = 0;
+	}
+
+	/* The length field in our CQE should be the full send buffer size. */
+	req->msg.recv_len = req->msg.send_len;
+
+	ret = __gnix_msg_recv_completion(req->gnix_ep, req);
+	if (ret != FI_SUCCESS)
+		GNIX_WARN(FI_LOG_EP_DATA,
+			  "__gnix_msg_recv_completion() failed: %d\n",
+			  ret);
+
+	return ret;
 }
 
-static int  __gnix_discard_request(struct gnix_fid_ep *ep,
-		struct gnix_fab_req *req,
-		void *addr,
-		size_t len,
-		void *context,
-		uint64_t tag,
-		uint64_t src_addr)
+static int __gnix_discard_request(struct gnix_fab_req *req)
 {
 	int ret = FI_SUCCESS;
 	int rendezvous = !!(req->msg.send_flags & GNIX_MSG_RENDEZVOUS);
 
+	/* The CQE should not contain a valid buffer. */
+	req->msg.recv_addr = 0;
+	req->msg.recv_len = 0;
+
 	GNIX_INFO(FI_LOG_EP_DATA, "discarding req=%p\n", req);
 	if (rendezvous) {
-		/* return a send completion so the sender knows the request/data
-		 * was sent, but discard the data locally
-		 */
```

Comment:
I think it just got removed with the rest of this block of code.  The code was redundant now that all the request fields are consistently set in _gnix_recv().  A similar comment remains just below this:

 "/* Complete rendezvous request, skipping data transfer. */"